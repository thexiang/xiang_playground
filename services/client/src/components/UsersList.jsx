import React, { useContext, useState } from "react";
import { Button, Input, InputNumber, Table } from 'antd';

import { UsersContext } from "context/usersContext";
import { AuthContext } from "context/authContext";
import { Card } from "./Card/Card";
import { TwoColumn } from "./Layout/TwoColumn";

const UsersList = () => {
	const [isEditing, setIsEditing] = useState(false);
	const [editId, setEditId] = useState(undefined);
	const [draftData, setDraftData] = useState({});

	const { dogs, deleteDog, updateDog, isLoadingDogs } = useContext(UsersContext);
	const { isAuthenticated } = useContext(AuthContext);

	const handleOnChange = (key, e) => {
		// InputNumber component's value returned directly from e
		const value = e?.target?.value ?? e;
		setDraftData({ ...draftData, [key]: value});
	}

	const handleClickEdit = (id) => {
		setIsEditing(!isEditing);
		setEditId(id);
	}

	const handleSave = (id) => {
		updateDog(id, draftData);
		setDraftData({});
		setIsEditing(false);
	}
	
	const columns = [
		{
			title: 'Id',
			dataIndex: 'id',
			key: 'id',
		},
		{
			title: 'Name',
			dataIndex: 'name',
			key: 'name',
			render: (_, record) => (
				<>
					{isEditing && editId === record.id 
						? <Input defaultValue={record.name} onChange={(e) => handleOnChange('name', e)} style={{ width: '100px' }} /> 
						: record.name
					}
				</>
			)
		},
		{
			title: 'Age',
			dataIndex: 'age',
			key: 'age',
			render: (_, record) => (
				<>
					{isEditing && editId === record.id 
						? <InputNumber defaultValue={record.age} onChange={(e) => handleOnChange('age', e)} style={{ width: '100px' }} /> 
						: record.age
					}
				</>
			)
		}
	];

	const authenticatedCols = [
		{
			title: 'Action',
			dataIndex: 'edit',
			key: 'edit',
			render: (_, record) => {
				const isEditingThis = isEditing && editId === record.id;
				const handleClick = isEditingThis ? () => handleSave(record.id) : () => handleClickEdit(record.id);
				return (
					<>
						<Button onClick={handleClick} style={{ marginRight: '10px' }}>
							{isEditingThis ? 'Save' : 'Edit'}
						</Button>
						<Button danger onClick={() => deleteDog(record.id)}>Delete</Button>
					</>
				)
			}
		}
	]

	const cols = isAuthenticated ? [...columns, ...authenticatedCols] : columns;
	
	return (
		<TwoColumn>
			<>
				<Card>
					<Table dataSource={dogs} columns={cols} loading={isLoadingDogs} />
				</Card>

				<Card>
					Hello
				</Card>
			</>
		</TwoColumn>
	);
};

export default UsersList;
