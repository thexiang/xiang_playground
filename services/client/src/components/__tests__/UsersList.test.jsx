import React from "react";
import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import UsersList from "../UsersList";

afterEach(cleanup);

const users = [
  {
    email: "nazumu@gmail.com",
    id: 1,
    username: "miya",
  },
  {
    email: "miya@mnazu.org",
    id: 2,
    username: "miyanazu",
  },
];

it("renders a username", () => {
  const { getByText } = render(
    <UsersList
      users={users}
      removeUser={() => true}
      isAuthenticated={() => true}
    />
  );
  expect(getByText("miya")).toHaveClass("username");
  expect(getByText("miyanazu")).toHaveClass("username");
});

it("renders", () => {
  const { asFragment } = render(
    <UsersList
      users={users}
      removeUser={() => true}
      isAuthenticated={() => false}
    />
  );
  expect(asFragment()).toMatchSnapshot();
});

it("renders when authenticated", () => {
  const { asFragment } = render(
    <UsersList
      users={users}
      removeUser={() => true}
      isAuthenticated={() => true}
    />
  );
  expect(asFragment()).toMatchSnapshot();
});
