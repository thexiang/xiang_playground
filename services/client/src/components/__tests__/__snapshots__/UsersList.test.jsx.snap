// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <div>
    <table
      class="table is-hoverable is-fullwidth"
    >
      <thead>
        <tr>
          <th>
            ID
          </th>
          <th>
            Email
          </th>
          <th>
            Username
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            1
          </td>
          <td>
            nazumu@gmail.com
          </td>
          <td
            class="username"
          >
            miya
          </td>
        </tr>
        <tr>
          <td>
            2
          </td>
          <td>
            miya@mnazu.org
          </td>
          <td
            class="username"
          >
            miyanazu
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`renders when authenticated 1`] = `
<DocumentFragment>
  <div>
    <table
      class="table is-hoverable is-fullwidth"
    >
      <thead>
        <tr>
          <th>
            ID
          </th>
          <th>
            Email
          </th>
          <th>
            Username
          </th>
          <th />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            1
          </td>
          <td>
            nazumu@gmail.com
          </td>
          <td
            class="username"
          >
            miya
          </td>
          <td>
            <button
              class="button is-danger is-small"
            >
              Delete
            </button>
          </td>
        </tr>
        <tr>
          <td>
            2
          </td>
          <td>
            miya@mnazu.org
          </td>
          <td
            class="username"
          >
            miyanazu
          </td>
          <td>
            <button
              class="button is-danger is-small"
            >
              Delete
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
